src/cpp/core/r_util/RSourceIndex.cpp

```diff
@@ -219,57 +219,76 @@ bool isMethodOrClassDefinition(const RToken& token)
 
 class IndexStatus
 {
+   
 public:
    
-   IndexStatus()
-      : braceLevel_(0),
-        parenLevel_(0),
-        bracketLevel_(0),
-        doubleBracketLevel_(0)
-   {}
+   IndexStatus(const RTokens& tokens)
+      : tokens_(tokens)
+   {
+   }
    
+   // The indexer maintains a vector of indices, recording
+   // the indices at which brackets were discovered. Tokens
+   // are popped off the stack as right brackets are discovered.
    void update(const RTokenCursor& cursor)
    {
       switch (cursor.type())
       {
       
-      case RToken::LBRACE: ++braceLevel_; break;
-      case RToken::LPAREN: ++parenLevel_; break;
-      case RToken::LBRACKET: ++bracketLevel_; break;
-      case RToken::LDBRACKET: ++doubleBracketLevel_; break;
-         
-      case RToken::RBRACE: --braceLevel_; break;
-      case RToken::RPAREN: --parenLevel_; break;
-      case RToken::RBRACKET: --bracketLevel_; break;
-      case RToken::RDBRACKET: --doubleBracketLevel_; break;
+      // Left brackets
+      case RToken::LPAREN:
+      case RToken::LBRACE:
+      case RToken::LBRACKET:
+      case RToken::LDBRACKET:
+      {
+         stack_.push_back(cursor.offset());
+         break;
+      }
+      
+      // Right brackets
+      case RToken::RBRACE:
+      case RToken::RPAREN:
+      case RToken::RBRACKET:
+      case RToken::RDBRACKET:
+      {
+         stack_.pop_back();
```

Comment:
What happens to this if the nesting is incorrect? e.g. with the code `{( })`, encountering the first `}` would seem to pop the `)` off the stack even though it doesn't match.

Reply:
Bad things! I've updated the PR to better handle this -- now, incompatible braces are just dropped. It might not be the best way of handling this (different methods have different tradeoffs) but I think on the balance it's okay to say that:

1. If your document is in a parsable state, then these introspections should work without issue;
2. If your document is not in a parsable state, we'll do our best (and definitely not crash) but some features may not work.

At least, with the default R code diagnostics, one will see little red Xs all over the editor to signal something is wrong (and so hopefully infer that is why some of these features might not be working)